enhancements/installer/component-selection: Suggest early annotation presubmits (#1174)
openshift/enhancements: enhancements/installer/component-selection.md

### How To Implement A New Capability

1. Select a name for the new capability, like `insights`.
2. Open a pull request adding `capability.openshift.io/name` annotations to the manifests that deliver the capability, like [this][bump-annotations].
    Some manifests should only be included if multiple capabilities are enabled, and in that case use the `+` delimiter between capability names, like `capability.openshift.io/name=insights+monitoring`.
3. In presubmit CI for the pull request, the cluster-version operator will [exclude manifests with unrecognized capabilities][cvo-exclude-unrecognized].
    Review presubmit results for your annotation pull request to ensure that:
    1. The capability is being completely removed, without leaving dangling resources that you forgot to annotate.
    2. CI suites which you expected to pass continue to pass.
        You may find that some test-cases assume or require your capabilities presence, and they may need to grow logic to skip or alter the test conditions when your capability is not installed (like [this][bump-origin]).

>     If your annotation addition spands multiple pull requests, either because the manifests being annotated span multiple repositories or because you also need to make test suite adjustments in other repositories), you may be able to [use cluster-bot][cluster-bot] to run tests on a release assembling multiple in-flight pull requests.

Reviewer: ```suggestion
    If your annotation addition spans multiple pull requests, either because the manifests being annotated span multiple repositories or because you also need to make test suite adjustments in other repositories), you may be able to [use cluster-bot][cluster-bot] to run tests on a release assembling multiple in-flight pull requests.
```

Author: Fixed in 40652c8d -> 03ed3ad7.

4. Introduce the new capablity name in the openshift/api repo, like [this][bump-api].
    1. If no [versioned capability set](https://github.com/openshift/api/blob/8324d657dee1d594a8a7768e5569fea6a8f887a9/config/v1/types_cluster_version.go#L257-L291) exists for the current OCP version under development, introduce one as part of your pull request.
5. Bump the openshift/api vendored dependency in the openshift/cluster-version-operator repo, like [this][bump-cvo-vendor], so the cluster-version operator will understand the new annotation and its `vCurrent` inclusion status.

    At this point, clusters installed with the `None` set will nominally not include the new capability, but if you had un-annotated manifests in previous versions, those will still be installed.
    Prereleases at this point may need to document this behavior ("this capability does not yet disable anything"), depending on the expected level of interest.

> 6. Land your annotation pull request(s), so asking to disable the new capability will actually work.

Reviewer: ```suggestion
6. Land your annotation pull request(s), so that the capability related resources will be ignored when the new capability is disabled
```

> 7. Bump the openshift/api vendored dependency in the openshift/installer repo, like [this][bump-installer-vendor], to allow folks to request the new capability via `additionalEnabledCapabilities` without the installer rejecting the unrecognzed capability name.

Reviewer: ```suggestion
7. Bump the openshift/api vendored dependency in the openshift/installer repo, like [this][bump-installer-vendor], to allow folks to request the new capability via `additionalEnabledCapabilities` without the installer rejecting the unrecognized capability name.
```

Reviewer: 8. check the e2e-no-caps periodic to confirm it is still green after your changes merge (any time after step 6)

Author: Fixed in 40652c8d -> 03ed3ad7.
